loadgen/test_settings_internal.cc

```diff
@@ -245,4 +249,122 @@ void TestSettingsInternal::LogSummary(AsyncSummary &summary) const {
 }
 
 }  // namespace loadgen
+
+
+int TestSettings::FromConfig(const std::string &path, const std::string &model,
+                             const std::string &scenario) {
+  std::map<std::string, std::string> kv;
+
+  // lookup key/value pairs from config
+  auto lookupkv = [&](const std::string &model, const std::string &scenario,
+                      const std::string &key, size_t* val_l, double* val_d, double multiplier=1.0) {
+    std::map<std::string, std::string>::iterator it;
+    std::string found;
+    // lookup exact key first
+    it = kv.find(model + scenario + "." + key);
+    if (it != kv.end()) {
+      found = it->second;
+    } else {
+    // lookup key with model wildcard
+    it = kv.find("*." + scenario + "." + key);
+      if (it != kv.end()) {
+        found = it->second;
+      } else {
+        return false;
+      }
+    }
+    // if we get here, found will be set
+    if (val_l) *val_l = strtoul(found.c_str(), 0, 0) * int(multiplier);
+    if (val_d) *val_d = strtod(found.c_str(), 0) * multiplier;
+    return true;
+  };
+
+  // dirt simple config parser
+  std::ifstream fss(path);
+  std::string line;
+  int line_nr = 0;
+  int errors = 0;
+  if (!fss.is_open()) {
+    LogDetail([p = path](AsyncDetail & detail) {
+      detail.Error("can't open ", p);
+    });
+    return -ENOENT;
+  }
+  while (std::getline(fss, line)) {
+    line_nr++;
+    std::istringstream iss(line);
+    std::string s, k;
+    int state = 0;
+    while (iss >> s) {
+      if (s == "#" && state != 2) {
+        // done with this line
+        break;
+      }
+      if (state == 2) {
+        // got key and value
+        const char *start = s.c_str();
+        char *stop; 
+        (void)strtoul(start, &stop, 0);
+        if (start + s.size() == stop) {
+          kv[k] = s;
+          continue;
+        }
+        (void)strtod(start, &stop);
+        if (start + s.size() == stop) {
+          kv[k] = s;
+          continue;
+        }
+        errors++;
+        LogDetail([l = line_nr](AsyncDetail & detail) {
+          detail.Error("error in line ", l);
```

Comment:
We need to figure out the error reporting mechanism before the test starts, but could you make the error messages a more descriptive?

Reply:
done